changelogs/unreleased/gh-9078-roles.md

```diff
@@ -0,0 +1,3 @@
+## feature/config
```

Comment:
I think it's time to start writting documentation requests, at least for new features.

As for me, it would also make the API of the user defined code more clear.

I would like to discuss an error convention here. `cartridge` uses `nil, err` everywhere, but I doubt that we should inherit it. At least, we should consider a raising error as a legal error reporting way. OTOH, support of `nil, err` and/or `false, err` may be ocnvenient for porting cartridge role code to the new framework. I don't mind it.

Anyway, let's write down the API description somewhere (preferably in the documentation request).

Reply:
Moved part of commit-message to doc-request. As for error convention, haven't thought about it yet.

Reply:
I think that we should at least write down the current calling convention, including an error convention. An author of a role should know what he/she should use to report a success or a failure.

Reply:
My personal opinion how the API of the role should look, at least for now.

```lua
local M = {}

-- Validates configuration of the role.
--
-- Called on initial configuration apply at startup and on
-- configuration reload if the role is enabled for the given instance.
--
-- The cfg argument may have arbitrary user provided value,
-- including nil.
--
-- Must raise an error if the validation fail.
function M.validate(cfg)
    -- <...>
end

-- Applies the given configuration of the role.
--
-- Called on initial configuration apply at startup and on
-- configuration reload if the role is enabled for the given instance.
--
-- The cfg argument may have arbitrary user provided value,
-- including nil.
--
-- Must raise an error if the given configuration can't be appied.
function M.apply(cfg)
    -- <...>
end

-- Stops the role.
--
-- Called on configuration reload if the role was enabled before
-- and removed now from the list of roles of the given instance.
--
-- Should cancel all background fibers and clean up hold
-- resources.
--
-- Must raise an error if this action can't be performed.
function M.stop()
    -- <...>
end

return M
```

Key points:

* No `init()`, because I see no difference from just placing initialization code into the module scope.
* `validate` and `apply` without `_config` suffix, because I see no much sense in the suffix.
* All the function raise an error at failure. I looked at several cartridge roles (for vshard, crud, metrics, expirationd) and all ones, where `validate_config` is present, raise an error from `validate_config` instead of returning `nil, err` or `false, err`.

The key idea is to start from a clean API and add cartridge compatibility only where it is actually needed.

Reply:
Rewrote doc-request.

Reply:
Thank you!

I have a few small comments regarding the texts.

1. Please, update the pull request description.
2. > In this regard, there is no strict order in which roles are performed.

   Seems obsoleted.
3. The documentation request doesn't mention arguments of role defined functions anywhere. (How about including a role example in the form of the snippet above?)
4. > On each run, the role calls validate() and then apply().

   The functions are called, but not by the role.
5. > Roles can be defined for each instance.

   May be interpreted as 'only on the instance scope', I would add 'for a replicaset, a group or for the whole cluster'.
6. There is no description of the new instance/cluster schema keys: `roles` and `roles_cfg`.

Reply:
Thanks, fixed.